src/main/java/com/yelp/nrtsearch/server/luceneserver/field/properties/HasSortableValueParser.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright 2023 Yelp Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.yelp.nrtsearch.server.luceneserver.field.properties;
+
+import com.yelp.nrtsearch.server.grpc.SearchResponse.Hit.CompositeFieldValue;
+import com.yelp.nrtsearch.server.grpc.SortType;
+
+/**
+ * * Trait interface for {@link com.yelp.nrtsearch.server.luceneserver.field.FieldDef} types,
+ * denoting that this field's sorted value shall be processed before return. We implicitly assume
+ * any class implement this interface is also {@link Sortable}.
+ */
+public interface HasSortableValueParser {
+
+  /**
+   * *
+   *
+   * @param querySortField the sort settings in query
+   * @param value the original sorted value
+   * @return the parsed sorted value in wrapped in {@link CompositeFieldValue}
+   */
+  CompositeFieldValue parseSortedValue(SortType querySortField, Object value);
```

Comment:
I might suggest a usage change here. Rather than this method doing the parsing, have it return an extractor method. Something like:

SortedValueExtractor getSortedValueExtractor(SortType querySortField);

Where `SortedValueExtractor` is a functional interface that takes a `SortField` and `Object` and produces a `CompositeFieldValue`. The default implementation could invoke (or just be) the existing `SortParser` method.

You can hold an array of extractors in the `SortContext` and don't need to worry about holding the `SortType` or `HasSortableValueParser` after parsing. This also lets you catch parsing errors, like invalid units, up front rather than the end of query execution.

Reply:
I thought about this extractor idea before. And I don't know how worth it is to create and cache extractors for each sortedField. Feels like in most cases, we just need the default extractor, and it will be repeated over and over again. Unless we want to create extractor per fieldDef to avoid switch-case in current default extractor, I don't think it worth the work of refactoring.

Reply:
I think my biggest issue with doing things this way is that queries don't fast fail on malformed requests. If an invalid units string is provided, an exception is not thrown until most the work is done.

Reply:
Changed into sortValueExtractor